landingai/storage/snowflake.py

```diff
@@ -0,0 +1,113 @@
+    """
+
+    user: str
+    password: str
+    account: str
+
+    class Config:
+        env_file = ".env"
+        env_prefix = "SNOWFLAKE_"
+        case_sensitive = False
+
+
+class SnowflakeDBConfig(BaseSettings):
+    """Snowflake connection config.
+    It supports loading from environment variables or .env files.
+
+    The supported name of the environment variables are (case-insensitive):
+    1. SNOWFLAKE_WAREHOUSE
+    2. SNOWFLAKE_DATABASE
+    3. SNOWFLAKE_SCHEMA
+
+    Environment variables will always take priority over values loaded from a dotenv file.
+    """
+
+    warehouse: str
+    database: str
+    # NOTE: the name "schema" is reserved by pydantic, so we use "snowflake_schema" instead.
+    snowflake_schema: str = Field(..., env="SNOWFLAKE_SCHEMA")
+
+    class Config:
+        env_file = ".env"
+        env_prefix = "SNOWFLAKE_"
+        case_sensitive = False
+
+
+def save_remote_file_to_local(
+    remote_filename: str,
+    stage_name: str,
+    *,
+    local_output: Path = None,
+    credential: Optional[SnowflakeCredential] = None,
+    connection_config: Optional[SnowflakeDBConfig] = None,
+) -> Path:
+    """Save a file stored in Snowflake to local disk.
+    If local_output is not provided, a temporary directory will be created and used.
+    If credential or connection_config is not provided, it will read from environment variable or .env file instead.
+    """
+    url = get_snowflake_presigned_url(remote_filename, stage_name, credential=credential, connection_config=connection_config)
+    if local_output is None:
+        local_output = Path(tempfile.mkdtemp())
+    file_path = local_output / remote_filename
+    file_path.parent.mkdir(parents=True, exist_ok=True)
+    with open(str(file_path), "wb") as f:
+        data_bytes = read_file(url)
+        f.write(data_bytes)
+    _LOGGER.info(f"Saved file {remote_filename} to {file_path}")
+    return file_path
+
+
+def get_snowflake_presigned_url(
+    remote_filename: str,
+    stage_name: str,
+    *,
+    credential: Optional[SnowflakeCredential] = None,
+    connection_config: Optional[SnowflakeDBConfig] = None,
+) -> str:
+    """Get a presigned url for a file stored in Snowflake."""
+    if credential is None:
+        credential = SnowflakeCredential()
+    if connection_config is None:
+        connection_config = SnowflakeDBConfig()
+
+    ctx = snowflake.connector.connect(
+        user=credential.user,
+        password=credential.password,
+        account=credential.account,
+        warehouse=connection_config.warehouse,
+        database=connection_config.database,
+        schema=connection_config.snowflake_schema,
+    )
+    cur = ctx.cursor()
+    result = cur.execute(f"LIST @{stage_name}")
+    files = result.fetchall()
+    _LOGGER.debug(f"Files in stage {stage_name}: {files}")
+    result = cur.execute(
+        f"SELECT get_presigned_url(@{stage_name}, '{remote_filename}') as url"
+    )
+    result_url = result.fetchall()[0][0]
```

Comment:
line 111, maybe we should add some checking if the remote_file does not exist?

Reply:
good point, added a check in here, just in case.

interesting finding: snowflake returns a valid url even i provide a file that doesn't exists. So i also added a check in other place.